src/edu/washington/escience/myria/operator/Limit.java

```diff
@@ -38,23 +41,30 @@ public Limit(@Nonnull final Long limit, final Operator child) {
     this.limit = Objects.requireNonNull(limit, "limit");
     Preconditions.checkArgument(limit >= 0L, "limit must be non-negative");
     toEmit = this.limit;
+    done = false;
   }
 
   @Override
   protected TupleBatch fetchNextReady() throws DbException {
     Operator child = getChild();
-    for (TupleBatch tb = child.nextReady(); tb != null; tb = child.nextReady()) {
-      if (tb.numTuples() <= toEmit) {
-        toEmit -= tb.numTuples();
-        return tb;
-      } else if (toEmit > 0) {
-        tb = tb.prefix(Ints.checkedCast(toEmit));
-        toEmit = 0;
-        return tb;
+    if (done) {
+      return null;
+    } else {
+      TupleBatch tb = child.nextReady();
+      TupleBatch result = null;
+      if (tb != null) {
+        if (tb.numTuples() <= toEmit) {
+          toEmit -= tb.numTuples();
+          result = tb;
+        } else if (toEmit > 0) {
```

Comment:
This logic is broken if `toEmit == 0` (implying that `tb.numTuples() == toEmit` on the previous call). In that case, you will return `null` on this and all subsequent calls until the child is exhausted, just like the previous `Limit` implementation. You need to close the child if `toEmit == 0` and return `null`. Also, please write a unit test which fails under this condition with the current implementation (just use a `TupleSource` containing as many tuples as the limit).